src/tools/illink/src/ILLink.Tasks/ILLink.Tasks.csproj

```
@@ -78,4 +78,24 @@
     </ProjectReference>
   </ItemGroup>
 
+  <Target Name="GenerateBuildMetadata" BeforeTargets="BeforeBuild"
+          Inputs="$(MSBuildProjectFile)"
+          Outputs="$(IntermediateOutputPath)Generated_BuildMetadata.cs">
+
+      <WriteLinesToFile
+        File="$(IntermediateOutputPath)Generated_BuildMetadata.cs"
+        Lines="// &lt;autogenerated /&gt;
+
+internal static class BuildMetadata
+{
+    public const string ILLinkImplAssemblyVersion = &quot;$(NetCoreAppToolCurrentVersion)&quot;%3B
+}
+  "
+        Overwrite="true" />
+
+      <ItemGroup>
+        <Compile Include="$(IntermediateOutputPath)Generated_BuildMetadata.cs" />
+      </ItemGroup>
```

nit: Add FileWrites so that the Clean target cleans this file:

```suggestion
        <FileWrites Include="$(IntermediateOutputPath)Generated_BuildMetadata.cs" />
      </ItemGroup>
```